File: dart_sample/lib/demos/material/material_switch_demo_page.dart
import 'package:material_ui/material_ui.dart';

class MaterialSwitchDemoPage extends StatefulWidget {
  const MaterialSwitchDemoPage({super.key});

  @override
  State<MaterialSwitchDemoPage> createState() => _MaterialSwitchDemoPageState();
}

class _MaterialSwitchDemoPageState extends State<MaterialSwitchDemoPage> {
  static const List<TargetPlatform> _platforms = <TargetPlatform>[
    TargetPlatform.android,
    TargetPlatform.iOS,
    TargetPlatform.macOS,
  ];

  bool _useMaterial3 = true;
  int _platformIndex = 0;
  bool _plain = true;
  bool _colored = false;
  bool _iconed = true;
  bool _outlined = false;
  bool _adaptive = false;

  @override
  Widget build(BuildContext context) {
    final ThemeData baseTheme = Theme.of(context);
    final TargetPlatform platform = _platforms[_platformIndex];
    final ThemeData pageTheme = baseTheme.copyWith(
      useMaterial3: _useMaterial3,
      platform: platform,
    );

    return Theme(
      data: pageTheme,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: <Widget>[
          const Text('Switch', style: TextStyle(fontSize: 20)),
          const SizedBox(height: 8),
          const Text(
            'M2/M3 tokens, thumb icons, track outlines and Switch.adaptive per platform.',
            style: TextStyle(fontSize: 14, color: Color(0x8A000000)),
          ),
          const SizedBox(height: 10),
          Row(
            children: <Widget>[
              _buildControlButton(
                label: _useMaterial3 ? 'M3' : 'M2',
                onTap: () => setState(() => _useMaterial3 = !_useMaterial3),
                width: 80,
                background: const Color(0xFFE9F0FF),
              ),
              const SizedBox(width: 8),
              _buildControlButton(
                label: platform.name,
                onTap: () => setState(
                  () => _platformIndex = (_platformIndex + 1) % _platforms.length,
                ),
                width: 112,
                background: const Color(0xFFEAF6F7),
              ),
            ],
          ),
          const SizedBox(height: 8),
          Text(
            'useMaterial3=${_useMaterial3 ? "true" : "false"}, '
            'platform=${platform.name}',
            style: const TextStyle(fontSize: 12, color: Color(0xFF607D8B)),
          ),
          const SizedBox(height: 10),
          Expanded(
            child: SingleChildScrollView(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.stretch,
                children: <Widget>[
                  _buildRow(
                    'Plain',
                    Switch(
                      value: _plain,
                      onChanged: (bool value) => setState(() => _plain = value),
                    ),
                  ),
                  const SizedBox(height: 14),
                  _buildRow(
                    'Custom colors',
                    Switch(
                      value: _colored,
                      onChanged: (bool value) => setState(() => _colored = value),
                      activeThumbColor: const Color(0xFFFFF8E1),
                      activeTrackColor: const Color(0xFF00695C),
                      inactiveThumbColor: const Color(0xFF8D6E63),
                      inactiveTrackColor: const Color(0xFFD7CCC8),
                    ),
                  ),
                  const SizedBox(height: 14),
                  _buildRow(
                    'Thumb icons',
                    Switch(
                      value: _iconed,
                      onChanged: (bool value) => setState(() => _iconed = value),
                      thumbIcon: WidgetStateProperty.resolveWith<Icon?>(
                        (Set<WidgetState> states) =>
                            states.contains(WidgetState.selected)
                            ? const Icon(Icons.check)
                            : const Icon(Icons.close),
                      ),
                    ),
                  ),
                  const SizedBox(height: 14),
                  _buildRow(
                    'Track outline',
                    Switch(
                      value: _outlined,
                      onChanged: (bool value) => setState(() => _outlined = value),
                      trackOutlineColor: const WidgetStatePropertyAll<Color?>(
                        Color(0xFF1565C0),
                      ),
                      trackOutlineWidth: const WidgetStatePropertyAll<double?>(3.0),
                    ),
                  ),
                  const SizedBox(height: 14),
                  _buildRow(
                    'Adaptive',
                    Switch.adaptive(
                      value: _adaptive,
                      onChanged: (bool value) => setState(() => _adaptive = value),
                    ),
                  ),
                  const SizedBox(height: 14),
                  _buildRow(
                    'Disabled (on)',
                    const Switch(value: true, onChanged: null),
                  ),
                  const SizedBox(height: 14),
                  _buildRow(
                    'Disabled (off)',
                    const Switch(value: false, onChanged: null),
                  ),
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildRow(String label, Widget control) {
    return Container(
      color: const Color(0xFFF7F9FC),
      padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 8),
      child: Row(
        children: <Widget>[
          SizedBox(
            width: 140,
            child: Text(
              label,
              style: const TextStyle(fontSize: 13, color: Colors.black),
            ),
          ),
          const SizedBox(width: 12),
          control,
        ],
      ),
    );
  }

  Widget _buildControlButton({
    required String label,
    required VoidCallback onTap,
    required double width,
    required Color background,
  }) {
    return SizedBox(
      width: width,
      child: TextButton(
        onPressed: onTap,
        style: TextButton.styleFrom(
          backgroundColor: background,
          foregroundColor: Colors.black,
          minimumSize: const Size(0, 36),
          padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
          shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(8)),
        ),
        child: Text(label, style: const TextStyle(fontSize: 12)),
      ),
    );
  }
}

File: src/Sample/Plumix.Sample/Demos/Material/MaterialSwitchDemoPage.cs
using System;
using Avalonia;
using Avalonia.Media;
using Plumix.Material;
using Plumix.Rendering;
using Plumix.UI;
using Plumix.Widgets;

// Dart parity source (reference): dart_sample/lib/demos/material/material_switch_demo_page.dart (exact sample parity)

namespace Plumix;

public sealed class MaterialSwitchDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new MaterialSwitchDemoPageState();
    }
}

internal sealed class MaterialSwitchDemoPageState : State
{
    private static readonly TargetPlatform[] Platforms =
    [
        TargetPlatform.Android,
        TargetPlatform.IOS,
        TargetPlatform.MacOS,
    ];

    private bool _useMaterial3 = true;
    private int _platformIndex;
    private bool _plain = true;
    private bool _colored;
    private bool _iconed = true;
    private bool _outlined;
    private bool _adaptive;

    public override Widget Build(BuildContext context)
    {
        ThemeData baseTheme = Theme.Of(context);
        TargetPlatform platform = Platforms[_platformIndex];
        ThemeData pageTheme = baseTheme with
        {
            UseMaterial3 = _useMaterial3,
            Platform = platform
        };

        return new Theme(
            data: pageTheme,
            child: new Column(
                crossAxisAlignment: CrossAxisAlignment.Stretch,
                spacing: 10,
                children:
                [
                    new Text("Switch", fontSize: 20, color: Colors.Black),
                    new Text(
                        "M2/M3 tokens, thumb icons, track outlines and Switch.adaptive per platform.",
                        fontSize: 14,
                        color: new Color(0x8A000000)),
                    new Row(
                        spacing: 8,
                        children:
                        [
                            BuildControlButton(
                                label: _useMaterial3 ? "M3" : "M2",
                                onTap: () => SetState(() => _useMaterial3 = !_useMaterial3),
                                width: 80,
                                background: new Color(0xFFE9F0FF)),
                            BuildControlButton(
                                label: platform.ToString(),
                                onTap: () => SetState(() =>
                                    _platformIndex = (_platformIndex + 1) % Platforms.Length),
                                width: 112,
                                background: new Color(0xFFEAF6F7)),
                        ]),
                    new Text(
                        $"useMaterial3={(_useMaterial3 ? "true" : "false")}, platform={platform}",
                        fontSize: 12,
                        color: new Color(0xFF607D8B)),
                    new Expanded(
                        child: new SingleChildScrollView(
                            child: new Column(
                                crossAxisAlignment: CrossAxisAlignment.Stretch,
                                spacing: 14,
                                children:
                                [
                                    BuildRow(
                                        "Plain",
                                        new Switch(
                                            value: _plain,
                                            onChanged: value => SetState(() => _plain = value))),
                                    BuildRow(
                                        "Custom colors",
                                        new Switch(
                                            value: _colored,
                                            onChanged: value => SetState(() => _colored = value),
                                            activeThumbColor: new Color(0xFFFFF8E1),
                                            activeTrackColor: new Color(0xFF00695C),
                                            inactiveThumbColor: new Color(0xFF8D6E63),
                                            inactiveTrackColor: new Color(0xFFD7CCC8))),
                                    BuildRow(
                                        "Thumb icons",
                                        new Switch(
                                            value: _iconed,
                                            onChanged: value => SetState(() => _iconed = value),
                                            thumbIcon: WidgetStateProperty<Icon?>.ResolveWith(
                                                states => states.Contains(WidgetState.Selected)
                                                    ? new Icon(Icons.Check)
                                                    : new Icon(Icons.Close)))),
                                    BuildRow(
                                        "Track outline",
                                        new Switch(
                                            value: _outlined,
                                            onChanged: value => SetState(() => _outlined = value),
                                            trackOutlineColor:
                                                WidgetStateProperty<Color?>.All(
                                                    new Color(0xFF1565C0)),
                                            trackOutlineWidth:
                                                WidgetStateProperty<double?>.All(3.0))),
                                    BuildRow(
                                        "Adaptive",
                                        Switch.Adaptive(
                                            value: _adaptive,
                                            onChanged: value => SetState(() => _adaptive = value))),
                                    BuildRow(
                                        "Disabled (on)",
                                        new Switch(value: true, onChanged: null)),
                                    BuildRow(
                                        "Disabled (off)",
                                        new Switch(value: false, onChanged: null)),
                                ]))),
                ]));
    }

    private static Widget BuildRow(string label, Widget control)
    {
        return new Container(
            color: new Color(0xFFF7F9FC),
            padding: new Thickness(12, 8),
            child: new Row(
                spacing: 12,
                children:
                [
                    new SizedBox(
                        width: 140,
                        child: new Text(label, fontSize: 13, color: Colors.Black)),
                    control,
                ]));
    }

    private Widget BuildControlButton(
        string label,
        Action onTap,
        double width,
        Color background)
    {
        return new SizedBox(
            width: width,
            child: new TextButton(
                onPressed: onTap,
                child: new Text(label, fontSize: 12),
                style: TextButton.StyleFrom(
                    foregroundColor: Colors.Black,
                    backgroundColor: background,
                    padding: new Thickness(10, 8),
                    minimumSize: new Size(64, 36),
                    shape: new RoundedRectangleBorder(
                        borderRadius: BorderRadius.Circular(8)))));
    }
}
